src/main/kotlin/com/amazon/opendistroforelasticsearch/indexmanagement/indexstatemanagement/elasticapi/ElasticExtensions.kt

```diff
@@ -84,3 +105,110 @@ fun getPolicyToTemplateMap(response: SearchResponse, xContentRegistry: NamedXCon
+        }
+    } catch (e: Exception) {
+        when (e) {
+            is IndexNotFoundException, is NoShardAvailableActionException -> {
+                log.error("Failed to get metadata because no index or shard not available")
+            }
+            else -> log.error("Failed to get metadata", e)
+        }
+
+        return null
+    }
+}
+
+/**
+ * multi-get metadata for indices
+ *
+ * @return list of metadata
+ */
+suspend fun Client.mgetManagedIndexMetadata(indices: List<Index>): List<ManagedIndexMetaData?> {
+    log.debug("trying to get back metadata for indices ${indices.map { it.name }}")
+
+    if (indices.isEmpty()) return emptyList()
+
+    val mgetRequest = MultiGetRequest()
+    indices.forEach {
+        mgetRequest.add(MultiGetRequest.Item(
+            INDEX_MANAGEMENT_INDEX, it.uuid + "metadata").routing(it.uuid))
+    }
+    var mgetMetadataList = mutableListOf<ManagedIndexMetaData?>()
+    try {
+        val response: MultiGetResponse = this.suspendUntil { multiGet(mgetRequest, it) }
+        mgetMetadataList = mgetResponseToList(response)
+    } catch (e: ActionRequestValidationException) {
+        log.info("No documents to get back metadata, ${e.message}")
+    }
+    return mgetMetadataList
+}
+
+/**
+ * transform multi-get response to list for ManagedIndexMetaData
+ */
+fun mgetResponseToList(mgetResponse: MultiGetResponse): MutableList<ManagedIndexMetaData?> {
+    val mgetList = mutableListOf<ManagedIndexMetaData?>()
+    mgetResponse.responses.forEach {
+        if (it.response != null && !it.response.isSourceEmpty) {
+            val xcp = contentParser(it.response.sourceAsBytesRef)
+            mgetList.add(ManagedIndexMetaData.parseWithType(
+                    xcp, it.response.id, it.response.seqNo, it.response.primaryTerm))
+        } else {
+            mgetList.add(null)
+        }
+    }
+
+    return mgetList
+}
+
+fun buildMgetMetadataRequest(clusterState: ClusterState): MultiGetRequest {
+    val mgetMetadataRequest = MultiGetRequest()
+    clusterState.metadata.indices.map { it.value.index }.forEach {
+        mgetMetadataRequest.add(MultiGetRequest.Item(
+            INDEX_MANAGEMENT_INDEX, it.uuid + "metadata").routing(it.uuid))
+    }
+    return mgetMetadataRequest
+}
+
+// forIndex means saving to config index, distinguish from Explain and History,
+// which only show meaningful partial metadata
+@Suppress("ReturnCount")
+fun XContentBuilder.addObject(name: String, metadata: ToXContentFragment?, params: ToXContent.Params, forIndex: Boolean = false): XContentBuilder {
+    if (!forIndex) { // for explain and history
+        return if (metadata != null) {
+            return this.buildMetadata(name, metadata, params)
+        } else {
+            this // omit this field
+        }
+    }
+    // else: save to config index
+    if (metadata != null) {
+        return this.buildMetadata(name, metadata, params)
```

Comment:
This looks the same as line 197? What is the difference or why do you need this then? If logic is correct then looks like it could be simplified to something like
```
if (metadata != null) return this.buildMetadata(name, metadata, params)
return if (forIndex) nullField(name) else this
```

Reply:
changed